db/view/view.cc

```diff
@@ -1658,7 +1659,7 @@ get_view_natural_endpoint(
         return {};
     }
     auto replica = view_endpoints[base_it - base_endpoints.begin()];
-    return topology.get_node(replica).endpoint();
+    return view_topology.get_node(replica).endpoint();
```

Comment:
I don't understand why we have two different topologies. The whole idea behind effective_replication_map_ptr is that while you hold it, topology can only change in a compatible way. So why not have view_erm be base_erm? Why have two variables in the first place?

Reply:
Topology is an object inside an erm. Erms are per table (I guess for tablets, but they have to be at least per keyspace even for vnodes since they are result of application of a replication to a topology).

Reply:
I don't think they're per table, I think they're per keyspace. @bhalevy

Reply:
They are per-keyspace with vnodes, and per-table with tablets.

Reply:
@tgrabiec is there a way to get to a "super effective replication map" that will cover both a base table and its views?

Reply:
Let's make effective_replication_map per keyspace always.

Reply:
> They are per-keyspace with vnodes, and per-table with tablets.

Well, since each table has its own pointer to _erm it does not matter that it is calculated per keyspace. In practice it is per table.

Reply:
> > They are per-keyspace with vnodes, and per-table with tablets.
> 
> Well, since each table has its own pointer to _erm it does not matter that it is calculated per keyspace. In practice it is per table.

Unless you are saying we can assume that base and view have the same erm and remove the use of view erm from mutate_MV all together. But what about tablets then? Also the API is confusing. A table should hold a pointer to a keyspace object and this object would hold the erm ptr.

Reply:
> > They are per-keyspace with vnodes, and per-table with tablets.
> 
> Well, since each table has its own pointer to _erm it does not matter that it is calculated per keyspace. In practice it is per table.

Hmm the erm instance indeed seems to be per-keyspace.
And it has a container for the per-table tablet_map:s

```c++
future<> database::add_column_family(keyspace& ks, schema_ptr schema, column_family::config cfg, is_new_cf is_new) {
    schema = local_schema_registry().learn(schema);
    schema->registry_entry()->mark_synced();
    auto&& rs = ks.get_replication_strategy();
    locator::effective_replication_map_ptr erm;
    if (auto pt_rs = rs.maybe_as_per_table()) {
        erm = pt_rs->make_replication_map(schema->id(), _shared_token_metadata.get());
    } else {
        erm = ks.get_vnode_effective_replication_map();
    }
    // avoid self-reporting
    auto& sst_manager = get_sstables_manager(system_keyspace(is_system_table(*schema)));
    auto cf = make_lw_shared<column_family>(schema, std::move(cfg), ks.metadata()->get_storage_options_ptr(), _compaction_manager, sst_manager, *_cl_stats, _row_cache_tracker, erm);
```

Reply:
> > They are per-keyspace with vnodes, and per-table with tablets.
> 
> Well, since each table has its own pointer to _erm it does not matter that it is calculated per keyspace. In practice it is per table.

Each request should inherit its erm from the caller. The table erm is there for table-internal processes. If everything uses a random erm, we'll never have consistent topology.

Reply:
Avi, we actually discussed this when deciding on the strategy for a fix and couldn't come up with a case when a view erm and base table erm from different "generations" could conflict. The streaming algorithm ensures that previous and current topology are compatible. Do you have a specific scenario in mind in which different erms could lead to consistency violations?

Reply:
Guys, maybe I'm missing something here, why this MV code had these "base" and "view" maps is that the base and the view are different tables, and as such, have different replication for a specific token.  For example we have in the get_view_natural_endpoint() code:
```
    for (auto&& base_endpoint : base_erm->get_replicas(base_token)) {
       ...
    }
    for (auto&& view_endpoint : view_erm->get_replicas(view_token)) {
       ...
    }
```
Although the base and view have the same keyspace, because they are a different table, the get_replicas() function can return a different result for the same token!

So it seems the claim made above that "base_erm" and "view_erm" are the same thing because they have the same keyspace can't be right... What am I missing???

Reply:
By the way, commit 4505a86f46dfafa039246369c46140a714efadbd is the commit which introduced the view_erm. It solved a real bug. It wasn't some sort of race or lifetime related bug - it was a real correctness bug, and I have a test that reproduces it. The view_erm and base_erm objects are really different and return different results for get_natural_endpoints() calls (I guess these were later renamed get_replicas?? I must have missed that change).

Reply:
@nyh, as I mentioned in https://github.com/scylladb/scylladb/pull/18816#discussion_r1616218546, with tablets, the base and view have different tablet maps, but they can both live in the same erm (which is kept in the keyspace.
A reason to use it, instead of 2 erm's possibly from different token_metadata generations, is that they would be based on the same cluster topology.
Also, with native tablet truncate, we would stop and drain tablet transitions using the parent erm, and then truncate the base and its views using the tablet maps in that erm instance.  That should be serialized properly also with view building / updates.
Cc @nodep @tgrabiec

Reply:
> > > They are per-keyspace with vnodes, and per-table with tablets.
> > 
> > 
> > Well, since each table has its own pointer to _erm it does not matter that it is calculated per keyspace. In practice it is per table.
> 
> Each request should inherit its erm from the caller.
The called passes erms here. The problem is that this request involves multiple tables each one with its own erm.

> The table erm is there for table-internal processes.
Then why is there public getter for it?

> If everything uses a random erm, we'll never have consistent topology.
I do not think we are discussing a relevant problem here. The problem is that no matter if the base and the view have the same erm or not if you take them non atomically you may get different erms. So you either need to take them atomically or not assume that both erms have the same topology. The patch here takes the second approach.

Reply:
@bhalevy maybe the function of "ERM" changed since I wrote this code (although, it does still work). As I noted above, what I needed to query was `base_erm->get_replicas(base_token)` - the list of three replicas that currently hold this token. The result of this is **different** between different tables. So if this is not an "erm" and is rather a "tablet map that is only a part of the erm", then I guess this code needs two "tablet maps", not two "erm"s. I'd appreciate it if someone who understand how this ERM stuff evolved will change this code to use the right objects or just write terminology. But again, the current code does work (and if it breaks, there's a test that exercises this case of different tablets for the base and view, and would have caught such breakage).

Reply:
> @nyh, as I mentioned in https://github.com/scylladb/scylladb/pull/18816#discussion_r1616218546, with tablets, the base and view have different tablet maps, but they can both live in the same erm (which is kept in the keyspace.

I do not see it from the code you posted. The code gets per_table_replication_strategy and creates individual erm for each table base on the schema id. May be I read the code wrong. Yes, the topology will be the same for all erms if they were created from the same token metadata, but replication for each tablet table is different.

Reply:
> I have a test that reproduces it

Is it committed to the repo?